go/libraries/doltcore/sqle/statspro/listener.go

```diff
@@ -0,0 +1,268 @@
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/dsess"
+)
+
+var ErrStatsIssuerPaused = fmt.Errorf("stats issuer is paused")
+
+type listenerEvent uint16
+
+const (
+	leUnknown               = listenerEvent(iota)
+	leSwap    listenerEvent = 1 << 0
+	leStop    listenerEvent = 1 << 1
+	leGc      listenerEvent = 1 << 2
+	leFlush   listenerEvent = 1 << 3
+)
+
+func (sc *StatsController) signalListener(s listenerEvent) {
+	var root, keep *listenMsg
+	n := sc.listeners
+	for n != nil {
+		if (n.e|leStop)&s > 0 {
+			n.c <- s
+			close(n.c)
+		} else if root == nil {
+			root = n
+			keep = n
+		} else {
+			keep.n = n
+			keep = n
+		}
+		n = n.n
+	}
+	if keep != nil {
+		keep.n = nil
+	}
+	sc.listeners = root
+}
+
+func (sc *StatsController) newThreadCtx(ctx context.Context) context.Context {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	newCtx, cancel := context.WithCancel(ctx)
+	if sc.activeCtxCancel != nil {
+		sc.activeCtxCancel()
+	}
+	sc.signalListener(leStop)
+	sc.activeCtxCancel = cancel
+	return newCtx
+}
+
+type listenMsg struct {
+	e listenerEvent
+	c chan listenerEvent
+	n *listenMsg
+}
+
+func (sc *StatsController) addListener(e listenerEvent) (chan listenerEvent, error) {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	if sc.activeCtxCancel == nil {
+		return nil, ErrStatsIssuerPaused
+	}
+	l := &listenMsg{e: e, c: make(chan listenerEvent, 1)}
+	if sc.listeners != nil {
+		l.n = sc.listeners
+	}
+	sc.listeners = l
+	return l.c, nil
+}
+
+func (sc *StatsController) Stop() {
+	// xxx: do not pause |sq|, analyze jobs still need to run
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	if sc.activeCtxCancel != nil {
+		sc.activeCtxCancel()
+		sc.activeCtxCancel = nil
+	}
+	sc.signalListener(leStop)
+	return
+}
+
+// UpdateParams reads the environment variables and updates controller
+// parameters. If the queue is not started this will hang.
+func (sc *StatsController) UpdateParams() {
+	_, memOnly, _ := sql.SystemVariables.GetGlobal(dsess.DoltStatsMemoryOnly)
+	sc.SetMemOnly(memOnly.(int8) == 1)
+
+	_, gcEnabled, _ := sql.SystemVariables.GetGlobal(dsess.DoltStatsGCEnabled)
+	sc.SetEnableGc(gcEnabled.(int8) == 1)
+
+	typ, jobI, _ := sql.SystemVariables.GetGlobal(dsess.DoltStatsJobInterval)
+	_, gcI, _ := sql.SystemVariables.GetGlobal(dsess.DoltStatsGCInterval)
+
+	jobInterval, _, _ := typ.GetType().Convert(jobI)
+	gcInterval, _, _ := typ.GetType().Convert(gcI)
+
+	sc.SetTimers(
+		jobInterval.(int64)*int64(time.Millisecond),
+		gcInterval.(int64)*int64(time.Millisecond),
+	)
+}
+
+func (sc *StatsController) Restart() error {
+	select {
+	case <-sc.closed:
+		return fmt.Errorf("StatsController is closed")
+	default:
+	}
+
+	sc.sq.Start()
+	sc.UpdateParams()
+
+	done := make(chan struct{})
+	go func() {
+		ctx := sc.newThreadCtx(context.Background())
+		close(done)
+		err := sc.runWorker(ctx)
+		if err != nil {
+			sc.logger.Errorf("stats stopped: %s", err.Error())
+		}
+	}()
+	// only return after latestCtx updated
+	<-done
```

Comment:
The sequence above doesn't make it clear that the ctx is set meaningfully until after the close signal, should close come after runWorker?

Reply:
runWorker lives as long as we're collecting stats